File: backend/models/meal_log.py
from pydantic import BaseModel, Field
from datetime import date
from typing import Optional, Literal, Dict, Any

class MealLog(BaseModel):
    meal_date: date
    meal_type: Literal['breakfast', 'lunch', 'dinner', 'snack']
    food_items: str = Field(min_length=1, max_length=500)
    calories: float = Field(ge=0, le=10000)
    protein: Optional[float] = Field(default=None, ge=0, le=1000)
    carbs: Optional[float] = Field(default=None, ge=0, le=1000)
    fats: Optional[float] = Field(default=None, ge=0, le=1000)
    base_meal_data: Optional[Dict[str, Any]] = None
    servings: Optional[float] = None

class MealLogUpdate(BaseModel):
    food_items: Optional[str] = None
    meal_type: Optional[Literal['breakfast', 'lunch', 'dinner', 'snack']] = None
    calories: Optional[float] = Field(default=None, ge=0, le=10000)
    protein: Optional[float] = Field(default=None, ge=0, le=1000)
    carbs: Optional[float] = Field(default=None, ge=0, le=1000)
    fats: Optional[float] = Field(default=None, ge=0, le=1000)
    servings: Optional[float] = None

class ProgressQuery(BaseModel):
    start_date: date
    end_date: date

File: meal_planner_app/lib/models/meal_log.dart
class MealLog {
  final int? logId;
  final String mealDate;
  final String mealType;
  final String foodItems;
  final double calories;
  final double? protein;
  final double? carbs;
  final double? fats;
  
  MealLog({
    this.logId,
    required this.mealDate,
    required this.mealType,
    required this.foodItems,
    required this.calories,
    this.protein,
    this.carbs,
    this.fats,
  });
  
  factory MealLog.fromJson(Map<String, dynamic> json) {
    return MealLog(
      logId: json['log_id'],
      mealDate: json['meal_date'],
      mealType: json['meal_type'],
      foodItems: json['food_items'],
      calories: json['calories'].toDouble(),
      protein: json['protein']?.toDouble(),
      carbs: json['carbs']?.toDouble(),
      fats: json['fats']?.toDouble(),
    );
  }
  
  Map<String, dynamic> toJson() {
    return {
      'meal_date': mealDate,
      'meal_type': mealType,
      'food_items': foodItems,
      'calories': calories,
      'protein': protein,
      'carbs': carbs,
      'fats': fats,
    };
  }
}